File: c/src/ble_privacy.c
// SPDX-License-Identifier: MIT
// BLE tracking protection: rotating Service UUID + IRK-based Resolvable Private
// Addresses (RPA).
//
// Faithful mirror of src/AetherNet.Security/Privacy/BlePrivacy.cs. The window is
// encoded as a little-endian int64; every AetherNet SDK reproduces the vectors
// in fixtures/bleprivacy/vectors.json byte-for-byte.
//
// Crypto reuses the SDK's existing libsodium HMAC-SHA256 backend
// (aethernet_hmac_sha256 in src/security.c). The BLE "ah" function needs a
// single AES-128-ECB block, which libsodium does not expose (its only AES is
// AES-256-GCM), so the public-domain tiny-AES-c in c/vendor/tiny-aes provides
// AES_ECB_encrypt for that one block. The vendored AES is built AES128/ECB-only
// via compile definitions in c/CMakeLists.txt.

#include <string.h>

#include "aethernet/ble_privacy.h"
#include "aethernet/security.h"   /* aethernet_hmac_sha256 */

#include "aes.h"                  /* c/vendor/tiny-aes (AES-128 ECB single block) */

/* Encode a signed 64-bit window as 8 little-endian bytes (matches
 * BinaryPrimitives.WriteInt64LittleEndian in the C# reference). */
static void window_le64(int64_t window, uint8_t out[8]) {
    uint64_t u = (uint64_t)window;
    for (int i = 0; i < 8; i++) {
        out[i] = (uint8_t)(u & 0xFF);
        u >>= 8;
    }
}

/* BLE "ah" hash: AES-128-ECB(irk, 0^13 || prand[3]) → first 3 bytes.
 * Single-block ECB, no padding — exactly the C# Ah(). */
static void ble_ah(const uint8_t irk[16], const uint8_t prand[3], uint8_t out_hash[3]) {
    uint8_t block[16];
    memset(block, 0, sizeof(block));
    memcpy(block + 13, prand, 3);

    struct AES_ctx ctx;
    AES_init_ctx(&ctx, irk);
    AES_ECB_encrypt(&ctx, block); /* in place; block now holds the ciphertext */

    out_hash[0] = block[0];
    out_hash[1] = block[1];
    out_hash[2] = block[2];
}

int64_t aethernet_ble_window_for(int64_t unix_seconds) {
    return unix_seconds / AETHERNET_BLE_ROTATION_SECONDS;
}

bool aethernet_ble_service_uuid(const uint8_t *rotation_key,
                                size_t rotation_key_len,
                                int64_t window,
                                char *out_uuid,
                                size_t out_cap) {
    if (!rotation_key || !out_uuid || out_cap < AETHERNET_BLE_UUID_STR_SIZE) {
        return false;
    }

    uint8_t win[8];
    window_le64(window, win);

    uint8_t mac[32];
    if (!aethernet_hmac_sha256(rotation_key, rotation_key_len, win, sizeof(win), mac)) {
        return false;
    }

    // Format mac[0..16] as the canonical lowercase 8-4-4-4-12 UUID:
    // bytes 0-3, 4-5, 6-7, 8-9, 10-15 (matches C# FormatUuid).
    static const char HEX[] = "0123456789abcdef";
    // Byte group layout: number of bytes per hyphen-separated group.
    static const int groups[] = { 4, 2, 2, 2, 6 };

    size_t bi = 0;   // byte index into mac
    size_t oi = 0;   // char index into out_uuid
    for (int g = 0; g < 5; g++) {
        if (g > 0) {
            out_uuid[oi++] = '-';
        }
        for (int k = 0; k < groups[g]; k++) {
            uint8_t b = mac[bi++];
            out_uuid[oi++] = HEX[(b >> 4) & 0xF];
            out_uuid[oi++] = HEX[b & 0xF];
        }
    }
    out_uuid[oi] = '\0';   // oi == 36 here

    return true;
}

bool aethernet_ble_resolvable_address(const uint8_t *irk,
                                      int64_t window,
                                      uint8_t *out_rpa) {
    if (!irk || !out_rpa) {
        return false;
    }

    uint8_t win[8];
    window_le64(window, win);

    uint8_t mac[32];
    if (!aethernet_hmac_sha256(irk, AETHERNET_BLE_IRK_SIZE, win, sizeof(win), mac)) {
        return false;
    }

    uint8_t prand[3];
    prand[0] = mac[0];
    prand[1] = mac[1];
    prand[2] = mac[2];
    prand[0] = (uint8_t)((prand[0] & 0x3F) | 0x40); // RPA address-type bits (0b01)

    uint8_t hash[3];
    ble_ah(irk, prand, hash);

    // rpa = hash(3) || prand(3)
    out_rpa[0] = hash[0];
    out_rpa[1] = hash[1];
    out_rpa[2] = hash[2];
    out_rpa[3] = prand[0];
    out_rpa[4] = prand[1];
    out_rpa[5] = prand[2];

    return true;
}

bool aethernet_ble_resolve_address(const uint8_t *irk,
                                   size_t irk_len,
                                   const uint8_t *rpa,
                                   size_t rpa_len) {
    if (!irk || !rpa ||
        irk_len != AETHERNET_BLE_IRK_SIZE ||
        rpa_len != AETHERNET_BLE_RPA_SIZE) {
        return false;
    }

    const uint8_t *prand = rpa + 3;   // the RPA's prand is its last 3 bytes
    uint8_t hash[3];
    ble_ah(irk, prand, hash);

    return hash[0] == rpa[0] && hash[1] == rpa[1] && hash[2] == rpa[2];
}

File: rust/src/security/ble_privacy.rs
// SPDX-License-Identifier: MIT
//! Bluetooth-LE tracking protection: a rotating Service UUID and IRK-based
//! Resolvable Private Addresses (RPA), so a mesh node is discoverable by its
//! peers without exposing a stable, trackable Bluetooth fingerprint on the air.
//!
//! - The Service UUID rotates every 15 minutes, HMAC-SHA256-derived from a
//!   shared rotation key and the current time window. Every node in the same
//!   window derives the same UUID, so peers still find each other — but a
//!   passive scanner sees an identifier that changes and cannot be linked over
//!   time.
//! - The node's stable id is removed from the advertisement; a peer that holds
//!   the node's 128-bit Identity Resolving Key (IRK) resolves its rotating
//!   6-byte RPA instead (the BLE "ah" function).
//!
//! The window-based operations are deterministic and byte-identical across every
//! AetherNet SDK (verified against `fixtures/bleprivacy/vectors.json`). The time
//! window is encoded as a little-endian `i64`, matching the C# reference
//! `BlePrivacy` (`BinaryPrimitives.WriteInt64LittleEndian`).

use aes::Aes128;
use cipher::{BlockCipherEncrypt, KeyInit};
use hmac::{Hmac, Mac};
use sha2::Sha256;

type HmacSha256 = Hmac<Sha256>;

/// Rotation period in seconds (15 minutes).
pub const ROTATION_SECONDS: i64 = 900;

/// The rotation window index for a Unix-seconds timestamp.
pub fn window_for(unix_seconds: i64) -> i64 {
    unix_seconds / ROTATION_SECONDS
}

/// The window encoded as the 8-byte little-endian `i64` HMAC input — identical
/// across every port.
fn window_bytes(window: i64) -> [u8; 8] {
    window.to_le_bytes()
}

/// HMAC-SHA256(key, le64(window)) -> 32-byte tag.
fn hmac_window(key: &[u8], window: i64) -> [u8; 32] {
    // hmac 0.13 moved `new_from_slice` to the `KeyInit` trait (from `digest`).
    use hmac::digest::KeyInit;
    let mut mac = <HmacSha256 as KeyInit>::new_from_slice(key)
        .expect("HMAC accepts a key of any length");
    mac.update(&window_bytes(window));
    mac.finalize().into_bytes().into()
}

/// The rotating BLE Service UUID for a rotation key and time window. Every node
/// sharing the rotation key derives the same UUID within the window, enabling
/// mutual discovery with no static identifier on the air.
///
/// `mac = HMAC-SHA256(rotation_key, le64(window))`; the first 16 bytes are
/// formatted as a lowercase canonical UUID
/// `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` (bytes 0-3, 4-5, 6-7, 8-9, 10-15).
pub fn service_uuid(rotation_key: &[u8], window: i64) -> String {
    let mac = hmac_window(rotation_key, window);
    format_uuid(&mac[0..16])
}

/// A 6-byte Resolvable Private Address for a 16-byte IRK and time window:
/// `hash(3) || prand(3)`, where `prand` is HMAC-derived (with the RPA
/// address-type bits set) and `hash = AES-128(IRK, prand-block)`. Rotates every
/// window; only a peer holding the IRK can link successive addresses.
///
/// Returns `None` if `irk` is not exactly 16 bytes (mirrors the C# reference,
/// which throws on a bad IRK length).
pub fn resolvable_address(irk: &[u8], window: i64) -> Option<[u8; 6]> {
    if irk.len() != 16 {
        return None;
    }

    let tag = hmac_window(irk, window);
    let mut prand = [tag[0], tag[1], tag[2]];
    prand[0] = (prand[0] & 0x3F) | 0x40; // RPA address-type bits (0b01)

    let hash = ah(irk, &prand);

    let mut rpa = [0u8; 6];
    rpa[0..3].copy_from_slice(&hash);
    rpa[3..6].copy_from_slice(&prand);
    Some(rpa)
}

/// `true` if `rpa` was generated from `irk` — i.e. this node recognises the peer
/// behind the rotating address. Returns `false` for any wrong length
/// (`irk != 16` or `rpa != 6`) or a hash mismatch, never panicking.
pub fn resolve_address(irk: &[u8], rpa: &[u8]) -> bool {
    if irk.len() != 16 || rpa.len() != 6 {
        return false;
    }

    let prand = [rpa[3], rpa[4], rpa[5]];
    let hash = ah(irk, &prand);
    hash[..] == rpa[0..3]
}

/// BLE "ah" hash: `AES-128-ECB(irk, 0^13 || prand)`, keep the first 3 bytes.
///
/// A single 16-byte block, ECB, no padding. `irk` is assumed to be 16 bytes
/// (callers check); a wrong length would panic in `Aes128::new_from_slice`, but
/// every public entry point validates the IRK length before reaching here.
fn ah(irk: &[u8], prand: &[u8; 3]) -> [u8; 3] {
    let mut block = [0u8; 16];
    block[13..16].copy_from_slice(prand);

    let cipher = Aes128::new_from_slice(irk).expect("IRK must be 16 bytes for AES-128");
    cipher.encrypt_block((&mut block).into());

    [block[0], block[1], block[2]]
}

/// Lowercase canonical UUID from 16 bytes: `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx`
/// grouping bytes 0-3, 4-5, 6-7, 8-9, 10-15.
fn format_uuid(b: &[u8]) -> String {
    debug_assert_eq!(b.len(), 16);
    let mut s = String::with_capacity(36);
    for (i, byte) in b.iter().enumerate() {
        if i == 4 || i == 6 || i == 8 || i == 10 {
            s.push('-');
        }
        s.push_str(&format!("{:02x}", byte));
    }
    s
}
